File: src/main/java/com/zepben/vertxutils/routing/handlers/params/BodyType.java
/*
 * Copyright 2020 Zeppelin Bend Pty Ltd
 *
 * This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at https://mozilla.org/MPL/2.0/.
 */

package com.zepben.vertxutils.routing.handlers.params;

import com.zepben.annotations.EverythingIsNonnullByDefault;
import io.vertx.core.buffer.Buffer;
import io.vertx.core.json.DecodeException;
import io.vertx.core.json.JsonArray;
import io.vertx.core.json.JsonObject;

@SuppressWarnings("WeakerAccess")
@EverythingIsNonnullByDefault
public class BodyType {

    public static final RequestValueConverter<Buffer, JsonObject> JSON_OBJECT = RequestValueConverter.create(
        "json object",
        buffer -> {
            if (buffer.length() == 0)
                throw new ValueConversionException("unable to decode empty body");

            try {
                return new JsonObject(buffer.toString());
            } catch (DecodeException ex) {
                throw new ValueConversionException(ex.getMessage());
            }
        });

    public static final RequestValueConverter<Buffer, JsonArray> JSON_ARRAY = RequestValueConverter.create(
        "json array",
        buffer -> {
            if (buffer.length() == 0)
                throw new ValueConversionException("unable to decode empty body");

            try {
                return new JsonArray(buffer.toString());
            } catch (DecodeException ex) {
                throw new ValueConversionException(ex.getMessage());
            }
        });
}

File: src/main/java/com/zepben/vertxutils/routing/handlers/params/BodyType.kt
/*
 * Copyright 2026 Zeppelin Bend Pty Ltd
 *
 * This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at https://mozilla.org/MPL/2.0/.
 */
package com.zepben.vertxutils.routing.handlers.params

import io.vertx.core.json.DecodeException
import io.vertx.core.json.JsonArray
import io.vertx.core.json.JsonObject
import io.vertx.ext.web.RequestBody

object BodyType {

    val JSON_OBJECT: RequestValueConverter<RequestBody, JsonObject> =
        RequestValueConverter.create("json object") { body ->
            if (body.length() == 0)
                throw ValueConversionException("unable to decode empty body")

            try {
                JsonObject(body.asString())
            } catch (ex: DecodeException) {
                throw ValueConversionException(ex.message)
            }
        }

    val JSON_ARRAY: RequestValueConverter<RequestBody, JsonArray> =
        RequestValueConverter.create("json array") { body ->
            if (body.length() == 0)
                throw ValueConversionException("unable to decode empty body")

            try {
                JsonArray(body.asString())
            } catch (ex: DecodeException) {
                throw ValueConversionException(ex.message)
            }
        }

}
